src/AWS.Deploy.ServerMode.Client/ServerModeSession.cs

```diff
@@ -0,0 +1,152 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.\r
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+
+namespace AWS.Deploy.ServerMode.Client
+{
+    /// <summary>
+    /// Helper class that allows launching deployment tool in server mode.
+    /// It abstracts the server mode setup, CLI command execution and retries when desired port is unavailable.
+    /// </summary>
+    public interface IServerModeSession
+    {
+        /// <summary>
+        /// Start deployment tool in server mode.
+        /// It creates symmetric key and tries to setup the server mode.
+        /// It also handles the retries when a desired port is unavailable in the provided range.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <returns><see cref="ServerModeStartInfo"/> instance containing connection information.</returns>
+        /// <exception cref="InternalServerModeException">Throws when deployment tool server failed to start for un unknown reason.</exception>
+        /// <exception cref="PortUnavailableException">Throws when deployment tool server failed to start due to unavailability of free ports.</exception>
+        Task<ServerModeStartInfo> Start(CancellationToken cancellationToken);
+    }
+
+    public class ServerModeSession : IServerModeSession
+    {
+        private const int TCP_PORT_ERROR = -100;
+
+        private readonly int _startPort;
+        private readonly int _endPort;
+        private readonly bool _diagnosticLoggingEnabled;
+
+        public ServerModeSession(int startPort = 10000, int endPort = 10100, bool diagnosticLoggingEnabled = false)
```

Comment:
Recapping some earlier discussions, the Toolkit will need to explicitly specify the tool path (of the CLI), and optionally provide a path to a JSON that contains other parameters. The JSON would optionally configure the start/end port, and whether diagnostic logging is enabled.  

We want to avoid a case where the Toolkit has to generate a json at runtime, which is why the tool path is excluded from the json file.

Its okay for this change to take place in a follow up PR.